cpp/src/arrow/adapters/orc/adapter_util.cc

```diff
@@ -315,13 +342,462 @@ Status AppendBatch(const liborc::Type* type, liborc::ColumnVectorBatch* batch,
+  if (array.null_count() || incoming_mask) {
+    batch->hasNulls = true;
+  }
+  for (; *orc_offset < length && *arrow_offset < arrow_length;
+       (*orc_offset)++, (*arrow_offset)++) {
+    if (array.IsNull(*arrow_offset) ||
+        (incoming_mask && !(*incoming_mask)[*orc_offset])) {
+      batch->notNull[*orc_offset] = false;
+    } else {
+      batch->notNull[*orc_offset] = true;
+      const uint8_t* data = fixed_size_binary_array.GetValue(*arrow_offset);
+      if (batch->data[*orc_offset]) delete batch->data[*orc_offset];
+      batch->data[*orc_offset] = new char[data_length];  // Do not include null
+      memcpy(batch->data[*orc_offset], data, data_length);
+      batch->length[*orc_offset] = data_length;
+    }
+  }
+  batch->numElements = *orc_offset;
+  return arrow::Status::OK();
+}
+
+// If Arrow supports 256-bit decimals we can not support it unless ORC does it
+arrow::Status WriteDecimalBatch(liborc::ColumnVectorBatch* column_vector_batch,
+                                int64_t* arrow_offset, int64_t* orc_offset,
+                                const int64_t& length, const arrow::Array& array,
+                                const std::vector<bool>* incoming_mask) {
+  const arrow::Decimal128Array& decimal128_array(
+      checked_cast<const arrow::Decimal128Array&>(array));
+  auto batch = checked_cast<liborc::Decimal128VectorBatch*>(column_vector_batch);
+  // Arrow uses 128 bits for decimal type and in the future, 256 bits will also be
+  // supported.
+  int64_t arrow_length = array.length();
+  if (!arrow_length) {
+    return arrow::Status::OK();
+  }
+  if (array.null_count() || incoming_mask) {
+    batch->hasNulls = true;
+  }
+  for (; *orc_offset < length && *arrow_offset < arrow_length;
+       (*orc_offset)++, (*arrow_offset)++) {
+    if (array.IsNull(*arrow_offset) ||
+        (incoming_mask && !(*incoming_mask)[*orc_offset])) {
+      batch->notNull[*orc_offset] = false;
+    } else {
+      batch->notNull[*orc_offset] = true;
+      uint8_t* raw_int128 =
+          const_cast<uint8_t*>(decimal128_array.GetValue(*arrow_offset));
+      uint64_t* lower_bits = reinterpret_cast<uint64_t*>(raw_int128);
+      int64_t* higher_bits = reinterpret_cast<int64_t*>(raw_int128 + 8);
+      batch->values[*orc_offset] = liborc::Int128(*higher_bits, *lower_bits);
+    }
+  }
+  batch->numElements = *orc_offset;
+  return arrow::Status::OK();
+}
+
+arrow::Status WriteStructBatch(liborc::ColumnVectorBatch* column_vector_batch,
+                               int64_t* arrow_offset, int64_t* orc_offset,
+                               const int64_t& length, const arrow::Array& array,
+                               const std::vector<bool>* incoming_mask) {
+  const arrow::StructArray& struct_array(checked_cast<const arrow::StructArray&>(array));
+  auto batch = checked_cast<liborc::StructVectorBatch*>(column_vector_batch);
+  std::shared_ptr<std::vector<bool>> outgoing_mask;
```

Comment:
as noted above you should probably use Buffer here and copy the bitmap buffer appropriately from the underlying Array, then pass that through so you can very quickly update current nullability.